hudi-flink-writer/src/main/java/org/apache/hudi/operator/KeyedWriteProcessFunction.java

```diff
@@ -0,0 +1,164 @@
+  private Collector<Tuple3<String, List<WriteStatus>, Integer>> output;
+  private int indexOfThisSubtask;
+  private String latestInstant;
+  private boolean hasRecordsIn;
+
+  /**
+   * Job conf.
+   */
+  private HudiFlinkStreamer.Config cfg;
+  /**
+   * Serializable hadoop conf.
+   */
+  private SerializableConfiguration serializableHadoopConf;
+  /**
+   * HoodieWriteConfig.
+   */
+  private HoodieWriteConfig writeConfig;
+
+  /**
+   * Write Client.
+   */
+  private transient HoodieFlinkWriteClient writeClient;
+
+  @Override
+  public void open(Configuration parameters) throws Exception {
+    super.open(parameters);
+
+    indexOfThisSubtask = getRuntimeContext().getIndexOfThisSubtask();
+
+    cfg = (HudiFlinkStreamer.Config) getRuntimeContext().getExecutionConfig().getGlobalJobParameters();
+
+    // hadoopConf
+    serializableHadoopConf = new SerializableConfiguration(new org.apache.hadoop.conf.Configuration());
+    // HoodieWriteConfig
+    writeConfig = StreamerUtil.getHoodieClientConfig(cfg);
+
+    HoodieFlinkEngineContext context = new HoodieFlinkEngineContext(serializableHadoopConf, new FlinkTaskContextSupplier(getRuntimeContext()));
+
+    // writeClient
+    writeClient = new HoodieFlinkWriteClient<>(context, writeConfig);
+  }
+
+  @Override
+  public void snapshotState(FunctionSnapshotContext context) throws Exception {
+    // get latest requested instant
+    String commitType = cfg.tableType.equals(HoodieTableType.COPY_ON_WRITE.name()) ? HoodieTimeline.COMMIT_ACTION : HoodieTimeline.DELTA_COMMIT_ACTION;
+    List<String> latestInstants = writeClient.getInflightsAndRequestedInstants(commitType);
+    latestInstant = latestInstants.isEmpty() ? null : latestInstants.get(0);
+
+    if (output != null && latestInstant != null && records.size() > 0) {
+      hasRecordsIn = true;
+      String instantTimestamp = latestInstant;
+      LOG.info("Upsert records, subtask id = [{}]  checkpoint_id = [{}}] instant = [{}], record size = [{}]", indexOfThisSubtask, context.getCheckpointId(), instantTimestamp, records.size());
+
+      List<WriteStatus> writeStatus;
+      switch (cfg.operation) {
+        case INSERT:
+          writeStatus = writeClient.insert(records, instantTimestamp);
+          break;
+        case UPSERT:
+          writeStatus = writeClient.upsert(records, instantTimestamp);
+          break;
+        default:
+          throw new HoodieFlinkStreamerException("Unknown operation : " + cfg.operation);
+      }
+      output.collect(new Tuple3<>(instantTimestamp, writeStatus, indexOfThisSubtask));
+      records.clear();
+    }
+  }
+
+  @Override
+  public void initializeState(FunctionInitializationContext functionInitializationContext) throws Exception {
+    // no operation
+  }
+
+  @Override
+  public void processElement(HoodieRecord hoodieRecord, Context context, Collector<Tuple3<String, List<WriteStatus>, Integer>> collector) throws Exception {
+    records.add(hoodieRecord);
```

Comment:
I am thinking one thing: if the interval of the checkpoint is too long. If this buffer would cause OOM?